File: tests/fixtures/elixir/control-flow.ex
defmodule Fixtures.ControlFlow do
  def run(items, input, replacement) do
    {left, right} = pair()
    left.()
    right.()

    callback = fn -> helper() end
    callback.()
    callback = replacement
    callback.()

    case input do
      {:ok, value} when is_valid(value) -> value.()
      _ -> fallback()
    end
    value.()

    with {:ok, first} <- fetch(),
         second = prepare(first),
         true <- accept(second) do
      first.()
      second.()
    else
      {:error, reason} -> reason.()
    end
    first.()
    second.()
    reason.()

    for item <- items,
        prepared = prepare(item),
        accept(prepared) do
      item.()
      prepared.()
    end
    item.()
    prepared.()

    receive do
      {:message, payload} -> payload.()
    after
      0 -> timeout()
    end
    payload.()

    reference = &helper/1
    reference
  end

  def helper(value), do: value
end

File: tests/fixtures/scala/control-flow.scala
object ControlFlow:
  def run(
    entries: List[() => Unit],
    pairs: List[(() => Unit, () => Unit)],
    fallback: () => Unit,
  ): Unit =
    for
      entry <- entries
      prepared = prepare(entry)
      if accept(prepared)
    do
      entry()
      prepared()

    for (first, second) <- pairs do
      first()
      second()
    first()

    entry()
    prepared()

    entries.foreach { current =>
      val (left, right) = pair(current)
      left()
      right()
    }
    current()
    left()

    lookup() match
      case Some(found) if validate(found) => found()
      case _ => fallback()
    found()

    try risky()
    catch
      case failure: RuntimeException => failure.getMessage()
    failure.getMessage()

class Factory(build: () => Unit):
  val current = build()

def helper(): Unit = ()

val scala2Reference = helper _
val scala3Reference = helper

def timing(replacement: () => Unit, holder: Holder): Unit =
  before()
  val before = () => helper()
  before()

  var callback = () => helper()
  callback()
  callback = replacement
  callback()

  holder.callback = replacement
  holder.callback()
